[C10D] FlightRecorder don't compute duration inside dump()

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116909
* #116905
* #114817

Computing duration can lead to a hang due to calling cudaEventQuery when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang.  Watchdog may already hang due to cuda event queries, and
we have the monitor to kill and dump if watchdog hangs.

Perhaps we should use a separate thread for computing durations..

And this still doesn't fix test_trace_while_stuck?

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225

wconstab committed:
[C10D] FlightRecorder don't compute duration inside dump()

Computing duration can lead to a hang due to calling cudaEventQuery when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang.  Watchdog may already hang due to cuda event queries, and
we have the monitor to kill and dump if watchdog hangs.

Perhaps we should use a separate thread for computing durations..

And this still doesn't fix test_trace_while_stuck?

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -495,14 +495,26 @@ struct NCCLTraceBuffer {
   }
 
-  void retire_id(c10::optional<size_t> id) {
+  // Don't set compute_duration=True from inside dump API,
```
I think `retire_id` is only called in one place, and compute_duration will always be true. But dump_entries also calls getDurationFromFirstEvent, and that isn't prevented by this flag. So if we end up dumping an event that just became, done but hasn't been retired, then we still see the dumper thread try to calculate a duration. I was thinking that this change would add a flag to `update_state` so that when called from `dump_entries` it would not calculate the duration, but if called from retire_id it would.

wconstab replied:
oh - yes i intended to delete the call to `getDurationFromFirstEvent` from inside `update_state`, and have only retire_id calling duration.  Thanks for catching that.

wconstab committed:
Update on "[C10D] FlightRecorder don't compute duration inside dump()"


Computing duration can lead to a hang due to calling cudaEventQuery when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang.  Watchdog may already hang due to cuda event queries, and
we have the monitor to kill and dump if watchdog hangs.

Perhaps we should use a separate thread for computing durations..

And this still doesn't fix test_trace_while_stuck?

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] FlightRecorder don't compute duration inside dump()"


Computing duration can lead to a hang due to calling cudaEventQuery when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang.  Watchdog may already hang due to cuda event queries, and
we have the monitor to kill and dump if watchdog hangs.

Perhaps we should use a separate thread for computing durations..

And this still doesn't fix test_trace_while_stuck?

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -495,16 +492,58 @@ struct NCCLTraceBuffer {
+    // can dump(), which we never want to block.
+    c10::optional<float> duration = c10::nullopt;
+    if (can_compute_duration) {
+      duration = getDurationFromFirstEvent(*startEvents, *endEvents);
```
the other way to do this that i thought about, was to make it use the async-launcher pattern.  Unlike the other uses, this would happen frequently so i wouldn't want to make a new thread per duration request, we'd need some long lived worker thread.  I might do this later anyway, and have it also service other types of requests like we were discussing.  But maybe for now keep it simple?

wconstab commented:
folded into the base PR that adds duration_ms- https://github.com/pytorch/pytorch/pull/114817
